File: Chapter_4/src/main/scala/sparksql_operations.scala
//   import packages

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

//defining a singleton scala object containing the main method

object sparksql_operations {
  def main(args:Array[String]) {

    // check the files passed
    if (args.length != 1) {
      println("Please pass the correct files")
      sys.exit(1)
    }

    // loading the file path 
    val data_file = args(0)

    // instantiating a spark session
    val spark = SparkSession.builder.appName("sparksql").getOrCreate()

    //defining schema using DDL

    val my_schema = "date DATE, delay INT, distance INT, origin STRING, destination STRING"

    // ********************* METADATA******************

    // The US flight delays data set has five columns:
    // The date column contains a string like 02190925. When converted, this maps to
    // 02-19 09:25 am.
    // The delay column gives the delay in minutes between the scheduled and actual
    //  departure times. Early departures show negative numbers.
    //  The distance column gives the distance in miles from the origin airport to the
    //  destination airport.
    //  The origin column contains the origin IATA airport code.
    //  The destination column contains the destination IATA airport code.

    // loading the datafile into a dataframe

    // val df = spark.read.format("csv").schema(my_schema).load(data_file)

    // since the date column is loading NULL values, spark is not able to parse the data correctly
    // therefore, loading the date column as "STRING" first then transform to DATE
    
    val updated_schema = "date STRING, delay INT, distance INT, origin STRING, destination STRING"

    val df = spark.read.format("csv").option("header","true").schema(updated_schema).load(data_file)
    
     //display the dataframe using sparksql

    df.show()
    df.printSchema()

    // converting the datatype of "date" column from STRING to DATE format

    val updated_df = df.withColumn("new_date", to_timestamp(col("date"), "MMddHHmm"))
    updated_df.printSchema()

    // Converting dataframe to SQL view
    
    updated_df.createOrReplaceTempView("us_flight_departure_delays")

    // ************************perform SPARKSQL operations on view*************************

    // 1. find all flights whose distance is greater than 1,000 miles

    val flights   = spark.sql("""SELECT * 
                                FROM us_flight_departure_delays
                                WHERE distance >1000
                                ORDER BY distance DESC""").show(10)


    // 2 .find all flights between San Francisco (SFO) and Chicago (ORD) with at least a two-hour delay:
    
    val flights_bw_sfo_ord = spark.sql("""SELECT * 
                                          FROM us_flight_departure_delays
                                          WHERE origin = 'SFO' 
                                          AND destination = 'ORD'
                                          AND delay >= 120
                                          ORDER BY delay DESC""").show(10)


    // 3. find the months when the delays were most common

    val months_with_most_delays = spark.sql(""" SELECT MONTH(new_date) AS delayed_months,
                                                COUNT("*") AS total_delays
                                                FROM us_flight_departure_delays
                                                WHERE origin = 'SFO' 
                                                AND destination = 'ORD'
                                                AND delay >= 120
                                                GROUP BY MONTH(new_date)
                                                ORDER BY total_delays DESC """).show(10)












    //stop the spark session

    spark.stop()
  }
}

File: Chapter_4/pyspark/sparksql_operations.py
#import packages

import sys
from pyspark.sql import SparkSession
from pyspark.sql.functions import *
from pyspark.sql.types import *

# creating the entry point of the application

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Please pass the script file & datafile",file = sys.stderr)
        sys.exit(1)

    # load the datafile's path
    data_file = sys.argv[1]

    # instantiating spark session

    spark =  SparkSession.builder.appName("sparksql_operations").getOrCreate()

    # defining the schema using DDL statement

    schema = "date STRING, delay INT, distance INT, origin STRING, destination STRING"

    # load the datafile into a dataframe with reference to defined schema

    df = spark.read.format("csv").option("header","true").schema(schema).load(data_file)

    # display the dataframe & schema
    df.show(10)
    df.printSchema()

    # creating a temp view to perform SparkSQL operations

    df.createOrReplaceTempView("us_flight_departure_delays")

    # query the temp view, NOTE=> querying a temp view using sparksql returns a dataframe

    new_df = spark.sql("""SELECT *
                            FROM us_flight_departure_delays""")
    new_df.show(10)
    # **************************SPARK SQL OPERATIONS ON TEMPORARY VIEW****************************  

    # 1.find all flights whose distance is greater than 1,000 miles 
    
    long_distance_flights = spark.sql("""SELECT *
                                         FROM us_flight_departure_delays
                                         WHERE distance > 1000
                                         ORDER BY distance DESC""").show(10)

    # 2. find all flights between San Francisco (SFO) and Chicago (ORD) with at least a two-hour delay

    flights_with_delays = spark.sql("""SELECT *
                                       FROM us_flight_departure_delays
                                       WHERE origin = 'SFO' AND
                                       destination = 'ORD' AND
                                       delay >= 120
                                       ORDER BY delay DESC""").show(20)

    # 3. find the months when the delays were most common

    # changing the datatype of "date" column from STRING to TIMESTAMP
    # since, the datatype of a temporary view cannot be changed as temp view is just a logical pointer
    # to a dataframe
    # therefore changing the datatype in dataframe & then forming a new temp view

    updated_df = df.withColumn("updated_date", to_timestamp(col("date"),"MMddHHmm"))

    updated_df.createOrReplaceTempView("us_flight_departure_delays")

    # print schema to verify addition of new column with changed datatype
    updated_df.printSchema()
    
    months_with_most_delays = spark.sql("""SELECT MONTH(updated_date) AS months_with_most_delays,
                                           COUNT ("*") AS total_months
                                           FROM us_flight_departure_delays
                                           WHERE origin = 'SFO' AND
                                           destination = 'ORD' AND
                                           delay >= 120
                                           GROUP BY MONTH(updated_date)
                                           ORDER BY total_months DESC""").show()

    # 4. Label all flights regardless of their origin or destination with an indication
    # of delays they experienced. For example:; Very Long delay > 6 hours, Long delay 2-6 hours 

    delayed_flight_label = spark.sql("""SELECT *,
                                        CASE
                                            WHEN delay > 360 THEN 'Very Long Delay'
                                            WHEN delay > 120 AND delay < 360 THEN 'Long Delay'
                                            WHEN delay > 60 AND delay < 120 THEN 'Delay'
                                            WHEN delay > 0 AND delay < 60 THEN 'Tolerable Delay'
                                            WHEN delay = 0 THEN 'No Delay'
                                        ELSE
                                            'Early'
                                        END AS Flight_delays
                                        FROM us_flight_departure_delays
                                        ORDER BY delay DESC""").show(15)










    
    #stop spark session
    spark.stop()

